### PR TITLE
Include both primary and mirror hosts in initial pg_hba.conf

### PR DESCRIPTION
**Pull request by d**
Include both primary and mirror hosts in initial pg_hba.conf

This commit fixes the issue that a mirror created from gpinitsystem is
not prepared to facilitate an incremental recovery after it's promoted.

We ensure that the initial pg_hba.conf generated for a segment always
includes both the primary and the mirror hosts. This allows utility
connection from either peer (as is used by pg_rewind).

Co-authored-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>

### CODE REVIEW

**Comment by d** [comment]
To the reviewers, we pushed an amendment that fixes a small bug we found while working on this patch. See the behave test for a clue

**Comment by bhuvnesh2703** [comment]
This PR now has 2 commits:
Commit 1. Fix for the github issue: https://github.com/greenplum-db/gpdb/issues/8432
Commit 2. Some refactoring of existing code

File: gpMgmt/bin/lib/gpcreateseg.sh [hunk 1]
```diff
@@ -93,21 +93,32 @@ SET_VAR () {
...
+}
+
 PROCESS_QE () {
 
     LOG_MSG "[INFO][$INST_COUNT]:-Start Function $FUNCNAME"
     LOG_MSG "[INFO][$INST_COUNT]:-Processing segment $GP_HOSTADDRESS"
 
     # do primary stuff
-    if [ x"" = x"$COPY_FROM_PRIMARY_HOSTADDRESS" ]; then
+    if [ x"" = x"$PRIMARY_HOSTADDRESS" ]; then
```
Comment (ashwinstar) [inline gpMgmt/bin/lib/gpcreateseg.sh]: I am not seeing any common piece of code between primary and mirror in this function. It seems better to have separate function for mirror and move mirror code to that function. Then this function can be only for primary with no if/else required and will be simpler to follow along. As currently it has
```
if (primary)
{
}

If (mirror)
{
}
else
{
}
```

Comment (bhuvnesh2703) [reply gpMgmt/bin/lib/gpcreateseg.sh]: makes sense.

File: gpMgmt/bin/gpinitsystem [hunk 2]
```diff
@@ -1404,11 +1412,17 @@ CREATE_QES_PRIMARY () {
...
             else
                 HAS_MIRRORS_OPTION=no
+                MIRROR_HOST=""
             fi
 
-			MASTER_HOSTNAME=${MASTER_HOSTNAME} HBA_HOSTNAMES=${HBA_HOSTNAMES} $GPCREATESEG $FLAG $$ 1 $I IS_PRIMARY $INST_COUNT  $LOG_FILE $HEAP_CHECKSUM \
+
+
+      ((I_INDEX++))
+			MASTER_HOSTNAME=${MASTER_HOSTNAME} HBA_HOSTNAMES=${HBA_HOSTNAMES} $GPCREATESEG $FLAG $$ 1 $I IS_PRIMARY "${MIRROR_HOST}" $INST_COUNT  $LOG_FILE $HEAP_CHECKSUM \
```
[the remark this answers is not in the document]
Comment (bhuvnesh2703) [reply gpMgmt/bin/gpinitsystem]: now they are common.

File: gpMgmt/bin/lib/gpcreateseg.sh [hunk 3]
```diff
@@ -225,19 +239,21 @@ PROCESS_QE () {
         fi
 
         # add localhost
-        $TRUSTED_SHELL ${GP_HOSTADDRESS} "$ECHO host     all          all         localhost      trust >> ${GP_DIR}/$PG_HBA"
 
         if [ x"" = x"$COPY_FROM_PRIMARY_HOSTADDRESS" ]; then
+            $TRUSTED_SHELL ${GP_HOSTADDRESS} "$ECHO host     all          all         localhost      trust >> ${GP_DIR}/$PG_HBA"
```
[the remark this answers is not in the document]
Comment (bhuvnesh2703) [reply gpMgmt/bin/lib/gpcreateseg.sh]: now we do all the work at primary, and for mirror we just copy it over..

**Pull request merged**